crossdock/server/server.py

```diff
@@ -118,91 +147,104 @@ def update_span(span):
-            self.call_downstream_http(span, trace_request, trace_response, response_writer)
-        elif trace_request.downstream.transport == Transport.TCHANNEL:
-            if tchannel_supported:
-                self.real_call_downstream_tchannel(span, trace_request, trace_response,
-                                                   response_writer)
+            if trace_request and trace_request.downstream is not None:
+                call_downstream(span, trace_request, tr, response_writer)
             else:
-                self.call_downstream_tchannel(span, trace_request, trace_response,
-                                              response_writer)
-        else:
-            response_writer.finish()
-            raise UnknownTransportException("%s" % trace_request.downstream.transport)
-
-    def call_downstream_http(self, span, trace_request, trace_response, response_writer):
-
-        def handle_response(response):
-            tr = serializer.traceresponse_from_json(response.body)
-            if tr.notImplementedError:
                 response_writer.write(serializer.obj_to_json(tr))
-            else:
-                trace_response.downstream = tr
-                response_writer.write(serializer.obj_to_json(trace_response))
+                response_writer.finish()
+
+
+@tornado.gen.coroutine
+def call_downstream(span, trace_request, trace_response, response_writer):
+    # TODO stop passing span around like this
+    if trace_request.downstream.transport == Transport.HTTP:
+        resp = yield call_downstream_http(span, trace_request, trace_response, response_writer)
+    elif trace_request.downstream.transport == Transport.TCHANNEL:
+        resp = yield call_downstream_tchannel(trace_request, trace_response, response_writer)
+    else:
+        if response_writer:
             response_writer.finish()
-
-        downstream = trace_request.downstream
-        url = "http://%s:%s/join_trace" % (downstream.host, downstream.port)
-        body = serializer.downstream_to_json(downstream.downstream)
-
-        req = tornado.httpclient.HTTPRequest(url=url, method="POST",
-                                             headers={"Content-Type": "application/json"},
-                                             body=body)
-        http_client.before_http_request(tornado_http.TornadoRequestWrapper(request=req),
-                                        lambda: span)
-        client = tornado.httpclient.AsyncHTTPClient()
-        client.fetch(req, handle_response)
-
-    def call_downstream_tchannel(self, span, trace_request, trace_response, response_writer):
-
-        tr = tchannelNotImplResponse()
-        response_writer.write(serializer.obj_to_json(tr))
+        raise UnknownTransportException("%s" % trace_request.downstream.transport)
+    raise tornado.gen.Return(resp)
+
+
+@tornado.gen.coroutine
+def call_downstream_http(span, trace_request, trace_response, response_writer):
```

Comment:
I would rather separate responsibilities. The endpoint should be responsible for generating response data and writing it out to `response_writer`. There's no reason why `response_writer` should be exposed to anyone else, as it leads to this ugly branching in this function as it tries to do things it should not be responsible for. All it needs to do is
 * format the request to json
 * call downstream
 * format response back into object